src/fhirtypes/ElementDefinition.ts

```diff
@@ -1819,13 +1823,30 @@ export class ElementDefinition {
     ) {
       let newElements: ElementDefinition[] = [];
       if (this.contentReference) {
-        // TODO: Implement recursive constraints once the approach is better clarified.
-        // See: https://chat.fhir.org/#narrow/stream/179252-IG-creation/topic/Clarification.20on.20contentReference
-
         // Get the original resource JSON so we unfold unconstrained reference
         const type = this.structDef.type;
         const json = fisher.fishForFHIR(type, Type.Resource);
-        if (json) {
+        // contentReference elements will not contain a type field, so we must fish for the StructDef and
+        // check the differential
+        const profileJson = fisher.fishForFHIR(this.structDef.id, Type.Profile);
+        if (this.hasProfileElementExtension(profileJson)) {
+          const def = StructureDefinition.fromJSON(profileJson);
+          // Content references start with #, slice that off to id of referenced element
+          const contentRefId = this.getContentReferenceId();
+          const referencedElement = def.findElement(contentRefId);
+          newElements = referencedElement?.children().map(e => {
+            const eClone = e.clone();
+            eClone.id = eClone.id.replace(referencedElement.id, this.id);
+            eClone.structDef = this.structDef;
+            eClone.captureOriginal();
+            return eClone;
+          });
+          if (newElements.length > 0) {
+            // If we successfully unfolded, this element is no longer a content reference
+            this.type = referencedElement.type;
+            delete this.contentReference;
+          }
+        } else if (json) {
```

Comment:
As far as I can tell, this code all repeats the code below, with the only major difference being which `def` we are using. Would we be able to update this so that we don't have the repeated code?